railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,388 @@
+package ctrlutils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"github.com/kong/deck/file"
+	"github.com/kong/kubernetes-ingress-controller/pkg/sendconfig"
+	kicclientset "github.com/kong/kubernetes-ingress-controller/railgun/pkg/clientset"
+	"github.com/prometheus/common/log"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	knative "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	knativeversioned "knative.dev/networking/pkg/client/clientset/versioned"
+	knativeApis "knative.dev/pkg/apis"
+	"knative.dev/pkg/network"
+)
+
+// dedicated function that process ingress/customer resource status update after configuration is updated within kong.
+func PullConfigUpdate(kongConfig sendconfig.Kong, log logr.Logger, ctx context.Context, kubeConfig *rest.Config, stopCh <-chan struct{}) {
+	log.Info("Launching Ingress Status Update Thread.")
+	var wg sync.WaitGroup
+	for {
+		select {
+		case updateDone := <-kongConfig.ConfigDone:
+			log.Info("receive configuration information. Update ingress status %v \n", updateDone)
+			wg.Add(1)
+			go UpdateIngress(&updateDone, log, ctx, kubeConfig, &wg)
+			wg.Wait()
```

Comment:
In its current state, the use of the waitgroup doesn't seem to matter because we will add the routine, and then wait for routines which will block the (current) goroutine. As such, we could just do this:

```suggestion
			UpdateIngress(&updateDone, log, ctx, kubeConfig, &wg)
```

And drop the `sync.WaitGroup` altogether, related to above comments.

Reply:
Alternatively: I suppose we could rely on _timestamps_ in order to to make ordering decisions but I'm not exactly sure yet how that would look as far as how we track those timestamps?

Reply:
We dont any subroutine UpdateIngress blocker Main Thread int this case, which will generate performance issue for concurrency.

Reply:
Clarified this offline.